### PR TITLE
Leopoly_Phase1_002_Mini-Promise

### PR DESCRIPTION
**Pull request by humbletim**
... PR staging; description and test plan pending

A JS promise like mini library to assist with general threading issues and used 
heavily for the Leopoly integration

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5774/app_12035_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5774/HighFidelity-Beta-PR12035-7e1be2b17c9a1640aa1dc5fb314ecdcf589ab24a.dmg

**Comment by MiladNazeri** [comment]
Overall Leopoly PR roadmap:
https://docs.google.com/document/d/1GJC6qYHICYccIqdmaGgtt30Nj12Q3FtFmejns7biyks/edit
This is 2/9

Labeling Do not merge until code is reviewed for #'s 1-8

File: libraries/shared/src/shared/MiniPromises.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+#include "ReadWriteLockable.h"
+#include <memory>
+
+class MiniPromise : public QObject, public std::enable_shared_from_this<MiniPromise>, public ReadWriteLockable {
+    Q_OBJECT
+public:
+    using handlerFunction = std::function<void(QString error, QVariantMap result)>;
+    using successFunction = std::function<void(QVariantMap result)>;
+    using errorFunction = std::function<void(QString error)>;
+    using handlers = QVector<handlerFunction>;
```
Comment (AndrewMeadows) [inline libraries/shared/src/shared/MiniPromises.h]: I would expect type definitions like these to start with uppercase:  `handlerFunction` --> `HandlerFunction`.  I don't see a specific rule in our coding standard about it, but when I browse our codebase I find that is the convention, except sometimes (not always) when overriding the typename of a class that already has a lowercase name, like `using vec3 = glm::vec3;`.

Perhaps some other hifi devs have opinions about this?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5903/app_12035_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5903/HighFidelity-Beta-PR12035-e94103633705d80c7095bd298b5e3703a5a5f9d9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5903/HighFidelity-Beta-PR12035-e94103633705d80c7095bd298b5e3703a5a5f9d9.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6037/app_12035_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6037/HighFidelity-Beta-PR12035-fc41bcca5af66e995de46551e72298cb0bdf1449.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6037/HighFidelity-Beta-PR12035-fc41bcca5af66e995de46551e72298cb0bdf1449.dmg

File: libraries/shared/src/shared/MiniPromises.h [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+//     auto combinedCallback = [=](QString error, QVariantMap result) { .... }
+//
+// * Callback Handlers are automatically invoked on the right thread (the Promise's thread).
+// * Callbacks can be assigned anytime during a Promise's life and "do the right thing".
+//   - ie: for code clarity you can define success cases first or choose to maintain time order
+// * "Deferred" concept can be used to publish outcomes.
+// * "Promise" concept be used to subscribe to outcomes.
+//
+// See AssetScriptingInterface.cpp for some examples of using to simplify chained async results.
+
```
Comment (AndrewMeadows) [inline libraries/shared/src/shared/MiniPromises.h]: These files need a license header.  Either the normal Apache license most of our other files have or something compatible.

I don't see any problem with the code itself.  I don't yet fully grok how to use it but I assume something would click if I were to see an example.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6372/app_12035_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6372/HighFidelity-Beta-PR12035-744da485512c691ea110145a8bb3a809402cf57e.dmg

**Comment by humbletim** [comment]
merging latest commits (including apache 2.0 headers) into 003 PR #12036

**Pull request closed without merging**